EIPS/eip-custom_errors_for_erc_tokens.md

```diff
@@ -0,0 +1,504 @@
+
+Ethereum applications (DApps) and wallets have historically used [revert statements](https://docs.soliditylang.org/en/v0.8.16/control-structures.html#revert-statement) to show relevant failure data to the users, however, this EIP provides a standard list of errors designed to give at least the same relevant information, but in a structured and expected way.
+
+## Motivation
+
+Since the introduction of Solidity custom errors in [v0.8.4](https://github.com/ethereum/solidity/releases/tag/v0.8.4), these have provided a way to show failures in a more expresive way with dynamic arguments, while reducing deployment costs.
+
+At the moment of the release of custom errors, the standard tokens ([EIP-20](./eip-20.md), [EIP-721](./eip-721.md), [EIP-1155](./eip-1155.md)) were already in finalized state, so no error specification was included.
+
+An error specification will allow users to expect more consistent error messages across applications or testing environments, while exposing pertinent arguments and overall reducing the need of writing expensive revert strings in the deployment bytecode.
+
+The list of token standard errors is defined in the specification below.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+The provided interfaces with errors SHOULD conform with the [error grammar rules](#error-grammar-rules) described in the [rationale](#rationale) section.
+
+It's important to note that errors MAY be thrown in each scenario, but that's up to the implementers, unless a revert reason is specified in their corresponding EIPs.
+
+This EIP aims to provide a clear standard only in case an error will be thrown.
+
+### [EIP-20](./eip-20.md)
+
+#### `ERC20InvalidSender(address)`
+
+Indicates a failure with the token sender.
+
+##### Parameters
+
+- `address _sender`: The address of whose tokens are transferred from.
+
+Throw when the cause of failure is the sender in a transfer.
+
+- MUST NOT be thrown for approval operations.
+- MUST NOT be thrown for balance or allowance requirements.
+  - Use `ERC20InsufficientBalance` or `ERC20InsufficientAllowance` instead.
+- MUST be thrown for unintended transfers from the zero address.
+
+#### `ERC20InvalidReceiver(address)`
+
+Indicates a failure with the token receiver.
+
+##### Parameters
+
+- `address _receiver` The address of whose tokens are transferred to.
+
+Throw when the cause of failure is the receiver in a transfer.
+
+- MUST NOT be thrown for approval operations.
+- MUST be thrown for unintended transfers to the zero address.
+- MUST be thrown for unintended transfers to non-compatible addresses (eg. contract addresses).
+
+#### `ERC20InsufficientBalance(address, uint256, uint256)`
```

Comment:
I would list errors in order to most common to least common. So insufficient balance should be the first one, then insufficient allowance, then the others.

Reply:
I'd define clearly what `most common` means.

Overall, the errors are following this arbitrary order:

First, transfer related:
1. Sender (address)
2. Receiver (address)
3. Balance (number)

Then, approval related:
4. Approver (address)
5. Spender (address)
6. Approval (number)

If we're moving `Balance` to the very beginning, I'd move `Approval` to the number 4 for consistency.
And, if we break the consistency, I'd put a formal definition of what `common` means.

What do you think?

Reply:
I'm going for the second one